pkg/controller/volume/cache/actual_state_of_world.go

```diff
@@ -0,0 +1,319 @@
+	// Used to generate the volume plugin object, and passed to attach/detach
+	// methods.
+	spec *volume.Spec
+
+	// nodesAttachedTo is a map containing the set of nodes this volume has
+	// successfully been attached to. The key in this map is the name of the
+	// node and the value is a node object containing more information about
+	// the node.
+	nodesAttachedTo map[string]nodeAttachedTo
+}
+
+// The nodeAttachedTo object represents a node that .
+type nodeAttachedTo struct {
+	// name contains the name of this node.
+	name string
+
+	// safeToDetach indicates that this node/volume combo has been unmounted
+	// by the node and is safe to detach
+	safeToDetach bool
+
+	// detachRequestedTime used to capture the desire to detach this volume
+	detachRequestedTime time.Time
+}
+
+func (asw *actualStateOfWorld) AddVolumeNode(volumeSpec *volume.Spec, nodeName string) (string, error) {
+	asw.Lock()
+	defer asw.Unlock()
+
+	attachableVolumePlugin, err := asw.volumePluginMgr.FindAttachablePluginBySpec(volumeSpec)
+	if err != nil || attachableVolumePlugin == nil {
+		return "", fmt.Errorf(
+			"failed to get AttachablePlugin from volumeSpec for volume %q err=%v",
+			volumeSpec.Name(),
+			err)
+	}
+
+	volumeName, err := attachableVolumePlugin.GetUniqueVolumeName(volumeSpec)
+	if err != nil {
+		return "", fmt.Errorf(
+			"failed to GetUniqueVolumeName from AttachablePlugin for volumeSpec %q err=%v",
+			volumeSpec.Name(),
+			err)
+	}
+
+	volumeObj, volumeExists := asw.attachedVolumes[volumeName]
+	if !volumeExists {
+		volumeObj = attachedVolume{
+			name:            volumeName,
+			spec:            volumeSpec,
+			nodesAttachedTo: make(map[string]nodeAttachedTo),
+		}
+		asw.attachedVolumes[volumeName] = volumeObj
+	}
+
+	nodeObj, nodeExists := volumeObj.nodesAttachedTo[nodeName]
+	if !nodeExists || nodeObj.safeToDetach || !nodeObj.detachRequestedTime.IsZero() {
+		// Create object if it doesn't exist.
+		// Reset safeToDeatch and detachRequestedTime values if it does.
+		volumeObj.nodesAttachedTo[nodeName] = nodeAttachedTo{
+			name:                nodeName,
+			safeToDetach:        false,
+			detachRequestedTime: time.Time{},
+		}
+	}
+
+	return volumeName, nil
+}
+
+func (asw *actualStateOfWorld) MarkVolumeNodeSafeToDetach(
+	volumeName, nodeName string) error {
+	asw.Lock()
+	defer asw.Unlock()
+	volumeObj, volumeExists := asw.attachedVolumes[volumeName]
+	if !volumeExists {
+		return fmt.Errorf(
+			"failed to MarkVolumeNodeSafeToDetach(volumeName=%q, nodeName=%q) volumeName does not exist",
+			volumeName,
+			nodeName)
+	}
+
+	nodeObj, nodeExists := volumeObj.nodesAttachedTo[nodeName]
+	if !nodeExists {
+		return fmt.Errorf(
+			"failed to MarkVolumeNodeSafeToDetach(volumeName=%q, nodeName=%q) nodeName does not exist",
+			volumeName,
+			nodeName)
+	}
+
+	volumeObj.nodesAttachedTo[nodeName] = nodeAttachedTo{
+		name:                nodeName,
+		safeToDetach:        true,
+		detachRequestedTime: nodeObj.detachRequestedTime,
```

Comment:
Why override the object instead of modifying it ? This is okay here but if nodeAttachedTo gets expanded in the future this code will need to be updated to copy the new fields as well.